EOBS (TG, TN, TX, RR, HU, PP)

norlandrhagen opened the pull request:
Initial attempt at the EOBS 23.1e 0.1 degree dataset. Some feedback/help on the input and chunking would be greatly appreciated!

This is a gridded daily dataset with a 0.1 degree spatial resolution. For the filepattern, this dataset is split up into 8 separate files, one for each variable. Each netcdf file contains `25933` time steps and is roughly 68GB  (`ds.nbytes/1e9 = 68`). 

A few questions:
1. Is it possible to specify a recipe without a `ConcatDim`?
2. Any general strategies for specifying `inputs_per_chunk` with such large file sizes.


Thanks!

cisaacstern commented:
> Is it possible to specify a recipe without a ConcatDim?

Recipes require a ConcatDim, but if you don't need to concatenate anything, you could do something like:

```python
def make_filename(time, variable):
    return f"https://data.com/{variable}{time}.nc"

concat_dim = ConcatDim("time", keys=[""])
```

> Any general strategies for specifying inputs_per_chunk with such large file sizes.

`inputs_per_chunk` is for input files which are smaller than worker memory (i.e. smaller than 500 MB). `subset_inputs` is for input files which are larger than memory. You'll want to use the latter, and specify it such that:

```
INPUT_FILE_SIZE / subset_inputs <= 500 MB
```

So if your input files are 68 GB, then you'll want `subset_inputs = {"time": 136}` because `68 GB / 136 = 500 MB`. (The subsetting has to happen over the ConcatDim. It doesn't matter that you're not actually concatenating anything.)

rabernat commented:
> Recipes require a ConcatDim

We should change this so that this sort of workaround is not needed.

rabernat commented on recipes/terraclimate/meta.yml:
```diff
@@ -0,0 +1,26 @@
+title: "TerraClimate Monthly Surface Dataset"
+description: "High-resolution monthly climate and climatic water balance dataset"
+pangeo_forge_version: "0.5.0"
+pangeo_notebook_version: "2021.07.17"
+recipes:
+  - id: terraclimate
+    object: "recipe:recipe"
+provenance:
+  providers:
+    - name: "Climatology Lab"
+      description: "Climatology Lab at University of California Merced (http://www.climatologylab.org/)"
+      roles:
+        - producer
+        - licensor
+      url: http://www.climatologylab.org/terraclimate.html
+  license: "CC-0-1.0"
+maintainers:
+  - name: "Raphael Hagen"
+    orcid: "0000-0003-1994-1153"
+    github: norlandrhagen
+bakery:
+  id: "devseed.bakery.development.aws.us-west-2"
+  target: pangeo-forge-aws-bakery-flowcachebucketdasktest4-10neo67y7a924
+  resources:
+    memory: 4096
+    cpu: 1024
```
terraclimate files don't belong in this PR

rabernat commented on recipes/terraclimate/recipe.py:
```diff
@@ -0,0 +1,116 @@
+
+
+def preproc(ds):
+    """custom preprocessing function for terraclimate data"""
+
+    rename = {}
+
+    station_influence = ds.get("station_influence", None)
+
+    if station_influence is not None:
+        ds = ds.drop_vars("station_influence")
+
+    var = list(ds.data_vars)[0]
+
+    if var in rename_vars:
+        rename[var] = rename_vars[var]
+
+    if "day" in ds.coords:
+        rename["day"] = "time"
+
+    if station_influence is not None:
+        ds[f"{var}_station_influence"] = station_influence
+
+    with xr.set_options(keep_attrs=True):
+        ds[var] = apply_mask(var, ds[var])
+
+    if rename:
+        ds = ds.rename(rename)
+
+    return ds
+
+
+# Recipe Inputs
+recipe = XarrayZarrRecipe(
+    file_pattern=pattern,
+    target_chunks=target_chunks,
+    copy_input_to_local_file=True,
+    subset_inputs={"time": 24},
+    process_chunk=preproc,
+)
```
terraclimate files don't belong in this PR

rabernat commented on recipes/EOBS/recipe.py:
```diff
@@ -0,0 +1,27 @@
+from pangeo_forge_recipes.recipes import XarrayZarrRecipe
+
+# Filename Pattern Inputs
+target_chunks = {"lat": 705, "lon": 465}
```
What about time? Is the goal here to have the data be contiguous (unchunked) in time?

rabernat commented on recipes/EOBS/recipe.py:
```diff
@@ -0,0 +1,27 @@
+
+# Recipe Inputs
+recipe = XarrayZarrRecipe(
+    file_pattern=pattern, target_chunks=target_chunks, subset_inputs={"time": 25933}
```
What is the total size of the input datasets in the `time` dimension?

rabernat reviewed (commented):
Thanks a lot for another contribution @norlandrhagen!

norlandrhagen commented:
Thanks for the feedback @rabernat  and @cisaacstern.

Updated changes:
-Added a spoof time concat dim.
-Time chunking in `target_chunks` to create daily slices. I might be completely misunderstanding this! 
-Updated `subset_inputs` from the above suggestion. Since each file contains a single variable, I think this chunking should meet the ~500Mb rec.
-Rebased to remove terraclimate files.

cisaacstern commented:
>  Time chunking in target_chunks to create daily slices. I might be completely misunderstanding this!

Approximately what size are each one of these `target_chunks`?

A naive explanation: if the entire dataset is 100 GB covering 1000 days, and you make daily `target_chunks`, a single daily chunk would be 100 GB / 1000 days = 100 MB.

Ideally, you want each one of your target chunks to be about 100 MB.

norlandrhagen commented:
> Approximately what size are each one of these `target_chunks`?

For a `target_chunk` containing all 8 variables a slice is ~21MB. 


<details>
<summary>Dataset Slice</summary>

```
xarray.Dataset

    Dimensions:
        longitude: 705latitude: 465
    Coordinates:
        longitude
        (longitude)
        float64
        -24.95 -24.85 ... 45.35 45.45
        latitude
        (latitude)
        float64
        25.05 25.15 25.25 ... 71.35 71.45
        time
        ()
        datetime64[ns]
        1950-04-11
    Data variables: (8)
    Attributes:

    CDI :
        Climate Data Interface version 1.6.3 (http://code.zmaw.de/projects/cdi)
    NCO :
        netCDF Operators version 4.7.5 (Homepage = http://nco.sf.net, Code = http://github.com/nco/nco)
    CDO :
        Climate Data Operators version 1.6.3 (http://code.zmaw.de/projects/cdo)
    E-OBS_version :
        23.1e
    Conventions :
        CF-1.4
    References :
        http://surfobs.climate.copernicus.eu/dataaccess/access_eobs.php


 xdf.nbytes/1e6
`>>>` 20.99
```

</details>

So to reach the ~100Mb `target_chunk_size` from the current ~20Mb,  each chunk could be every 5 days instead of daily. 

From: `target_chunks = {"lat": 705, "lon": 465, "time": 25933}` to `target_chunks = {"lat": 705, "lon": 465, "time": 5187}`

I think this gets us to the target chunk size, but once again, may be off base. 

Thanks again for the help!

cisaacstern commented:
Nice! I'd actually like clarification on this too...

@rabernat, should chunks be 100 MB per time slice inclusive of all variables? Or should they rather be 100 MB per time slice of a single variable?

cisaacstern commented:
@norlandrhagen, thanks for working on this! The latest error is because this PR deletes `.pre-commit-config.yaml`. If you restore that file from the `master` branch into this PR, this error should be resolved.

Also, you'll need to update `meta.yaml` -> `bakery` section to use `"pangeo-ldeo-nsf-earthcube"`, and you can leave off the other config under that section, which we don't currently use. 🙏

norlandrhagen commented:
@cisaacstern, got it! Just updated the .yml's. 

When running the recipe locally, I'm running into a potential hdf5 issue. Wondering if you've seen it before. 


`AttributeError: 'NoneType' object has no attribute '_h5file'` 

full traceback: 
<details>

``` 
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=0, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/pp_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/pp_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Subsetting input according to time-0
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Combining inputs for chunk 'Index({DimIndex(name='time', index=0, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Storing dataset in /var/folders/mb/7d7yq_4j2qgdfm_j3j4tsyl40000gn/T/tmpojh8vz3d/nUn30IAH
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=0, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=5, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=5, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/hu_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/hu_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Subsetting input according to time-0
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Combining inputs for chunk 'Index({DimIndex(name='time', index=0, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=5, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Storing dataset in /var/folders/mb/7d7yq_4j2qgdfm_j3j4tsyl40000gn/T/tmpojh8vz3d/nUn30IAH
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Expanding target concat dim 'time' to size 25933
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=260, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>), DimIndex(name='variable', index=0, sequence_len=6, operation=<CombineOp.MERGE: 1>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=375, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=3, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=489, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>), DimIndex(name='variable', index=0, sequence_len=6, operation=<CombineOp.MERGE: 1>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=84, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='variable', index=1, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=198, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='time', index=313, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='variable', index=1, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=607, sequence_len=700, operation=<CombineOp.SUBSET: 3>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening inputs for chunk Index({DimIndex(name='variable', index=5, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>), DimIndex(name='time', index=22, sequence_len=700, operation=<CombineOp.SUBSET: 3>)})
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=0, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tg_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=3, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/rr_ens_mean_0.1deg_reg_v23.1e.nc'

pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=0, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tg_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/pp_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=1, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tn_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=4, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/pp_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=1, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tn_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.recipes.xarray_zarr - INFO - Opening input with Xarray Index({DimIndex(name='variable', index=5, sequence_len=6, operation=<CombineOp.MERGE: 1>), DimIndex(name='time', index=0, sequence_len=1, operation=<CombineOp.CONCAT: 2>)}): 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/hu_ens_mean_0.1deg_reg_v23.1e.nc'
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tg_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/rr_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tg_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/pp_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tn_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/pp_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tn_ens_mean_0.1deg_reg_v23.1e.nc' from cache
pangeo_forge_recipes.storage - INFO - Opening 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/hu_ens_mean_0.1deg_reg_v23.1e.nc' from cache

---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [6], in <cell line: 2>()
      1 delayed = recipe.to_dask()
----> 2 delayed.compute()

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/base.py:315, in DaskMethodsMixin.compute(self, **kwargs)
    291 def compute(self, **kwargs):
    292     """Compute this dask collection
    293 
    294     This turns a lazy Dask collection into its in-memory equivalent.
   (...)
    313     dask.base.compute
    314     """
--> 315     (result,) = compute(self, traverse=False, **kwargs)
    316     return result

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/base.py:603, in compute(traverse, optimize_graph, scheduler, get, *args, **kwargs)
    600     keys.append(x.__dask_keys__())
    601     postcomputes.append(x.__dask_postcompute__())
--> 603 results = schedule(dsk, keys, **kwargs)
    604 return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/threaded.py:89, in get(dsk, keys, cache, num_workers, pool, **kwargs)
     86     elif isinstance(pool, multiprocessing.pool.Pool):
     87         pool = MultiprocessingPoolExecutor(pool)
---> 89 results = get_async(
     90     pool.submit,
     91     pool._max_workers,
     92     dsk,
     93     keys,
     94     cache=cache,
     95     get_id=_thread_get_id,
     96     pack_exception=pack_exception,
     97     **kwargs,
     98 )
    100 # Cleanup pools associated to dead threads
    101 with pools_lock:

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/local.py:511, in get_async(submit, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, chunksize, **kwargs)
    509         _execute_task(task, data)  # Re-execute locally
    510     else:
--> 511         raise_exception(exc, tb)
    512 res, worker_id = loads(res_info)
    513 state["cache"][key] = res

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/local.py:319, in reraise(exc, tb)
    317 if exc.__traceback__ is not tb:
    318     raise exc.with_traceback(tb)
--> 319 raise exc

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/local.py:224, in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    222 try:
    223     task, data = loads(task_info)
--> 224     result = _execute_task(task, data)
    225     id = get_id()
    226     result = dumps((result, id))

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/optimization.py:990, in SubgraphCallable.__call__(self, *args)
    988 if not len(args) == len(self.inkeys):
    989     raise ValueError("Expected %d args, got %d" % (len(self.inkeys), len(args)))
--> 990 return core.get(self.dsk, self.outkey, dict(zip(self.inkeys, args)))

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/core.py:149, in get(dsk, out, cache)
    147 for key in toposort(dsk):
    148     task = dsk[key]
--> 149     result = _execute_task(task, cache)
    150     cache[key] = result
    151 result = _execute_task(out, cache)

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/dask/core.py:119, in _execute_task(arg, cache, dsk)
    115     func, args = arg[0], arg[1:]
    116     # Note: Don't assign the subtask results to a variable. numpy detects
    117     # temporaries by their reference count and can execute certain
    118     # operations in-place.
--> 119     return func(*(_execute_task(a, cache) for a in args))
    120 elif not ishashable(arg):
    121     return arg

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/pangeo_forge_recipes/executors/dask.py:16, in wrap_map_task.<locals>.wrapped(map_arg, config, *dependencies)
     15 def wrapped(map_arg, config, *dependencies):
---> 16     return function(map_arg, config=config)

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py:585, in store_chunk(chunk_key, config)
    582 if config.storage_config.target is None:
    583     raise ValueError("target has not been set.")
--> 585 with open_chunk(chunk_key, config=config) as ds_chunk:
    586     # writing a region means that all the variables MUST have concat_dim
    587     to_drop = [v for v in ds_chunk.variables if config.concat_dim not in ds_chunk[v].dims]
    588     ds_chunk = ds_chunk.drop_vars(to_drop)

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/contextlib.py:119, in _GeneratorContextManager.__enter__(self)
    117 del self.args, self.kwds, self.func
    118 try:
--> 119     return next(self.gen)
    120 except StopIteration:
    121     raise RuntimeError("generator didn't yield") from None

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py:396, in open_chunk(chunk_key, config)
    394 # need to open an unknown number of contexts at the same time
    395 with ExitStack() as stack:
--> 396     dsets = [stack.enter_context(open_input(input_key, config=config)) for input_key in inputs]
    398     # subset before chunking; hopefully lazy
    399     subset_dims = [dim_idx for dim_idx in chunk_key if dim_idx.operation == CombineOp.SUBSET]

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py:396, in <listcomp>(.0)
    394 # need to open an unknown number of contexts at the same time
    395 with ExitStack() as stack:
--> 396     dsets = [stack.enter_context(open_input(input_key, config=config)) for input_key in inputs]
    398     # subset before chunking; hopefully lazy
    399     subset_dims = [dim_idx for dim_idx in chunk_key if dim_idx.operation == CombineOp.SUBSET]

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/contextlib.py:448, in _BaseExitStack.enter_context(self, cm)
    446 _cm_type = type(cm)
    447 _exit = _cm_type.__exit__
--> 448 result = _cm_type.__enter__(cm)
    449 self._push_cm_exit(cm, _exit)
    450 return result

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/contextlib.py:119, in _GeneratorContextManager.__enter__(self)
    117 del self.args, self.kwds, self.func
    118 try:
--> 119     return next(self.gen)
    120 except StopIteration:
    121     raise RuntimeError("generator didn't yield") from None

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/pangeo_forge_recipes/recipes/xarray_zarr.py:338, in open_input(input_key, config)
    336 logger.debug(f"about to enter xr.open_dataset context on {f}")
    337 try:
--> 338     with xr.open_dataset(f, **kw) as ds:
    339         logger.debug("successfully opened dataset")
    340         ds = fix_scalar_attr_encoding(ds)

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/api.py:495, in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, backend_kwargs, *args, **kwargs)
    483 decoders = _resolve_decoders_kwargs(
    484     decode_cf,
    485     open_backend_dataset_parameters=backend.open_dataset_parameters,
   (...)
    491     decode_coords=decode_coords,
    492 )
    494 overwrite_encoded_chunks = kwargs.pop("overwrite_encoded_chunks", None)
--> 495 backend_ds = backend.open_dataset(
    496     filename_or_obj,
    497     drop_variables=drop_variables,
    498     **decoders,
    499     **kwargs,
    500 )
    501 ds = _dataset_from_backend_dataset(
    502     backend_ds,
    503     filename_or_obj,
   (...)
    510     **kwargs,
    511 )
    512 return ds

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/h5netcdf_.py:399, in H5netcdfBackendEntrypoint.open_dataset(self, filename_or_obj, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta, format, group, lock, invalid_netcdf, phony_dims, decode_vlen_strings)
    387 store = H5NetCDFStore.open(
    388     filename_or_obj,
    389     format=format,
   (...)
    394     decode_vlen_strings=decode_vlen_strings,
    395 )
    397 store_entrypoint = StoreBackendEntrypoint()
--> 399 ds = store_entrypoint.open_dataset(
    400     store,
    401     mask_and_scale=mask_and_scale,
    402     decode_times=decode_times,
    403     concat_characters=concat_characters,
    404     decode_coords=decode_coords,
    405     drop_variables=drop_variables,
    406     use_cftime=use_cftime,
    407     decode_timedelta=decode_timedelta,
    408 )
    409 return ds

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/store.py:24, in StoreBackendEntrypoint.open_dataset(self, store, mask_and_scale, decode_times, concat_characters, decode_coords, drop_variables, use_cftime, decode_timedelta)
     12 def open_dataset(
     13     self,
     14     store,
   (...)
     22     decode_timedelta=None,
     23 ):
---> 24     vars, attrs = store.load()
     25     encoding = store.get_encoding()
     27     vars, attrs, coord_names = conventions.decode_cf_variables(
     28         vars,
     29         attrs,
   (...)
     36         decode_timedelta=decode_timedelta,
     37     )

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/common.py:123, in AbstractDataStore.load(self)
    101 def load(self):
    102     """
    103     This loads the variables and attributes simultaneously.
    104     A centralized loading function makes it easier to create
   (...)
    120     are requested, so care should be taken to make sure its fast.
    121     """
    122     variables = FrozenDict(
--> 123         (_decode_variable_name(k), v) for k, v in self.get_variables().items()
    124     )
    125     attributes = FrozenDict(self.get_attrs())
    126     return variables, attributes

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/h5netcdf_.py:230, in H5NetCDFStore.get_variables(self)
    229 def get_variables(self):
--> 230     return FrozenDict(
    231         (k, self.open_store_variable(k, v)) for k, v in self.ds.variables.items()
    232     )

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/core/utils.py:473, in FrozenDict(*args, **kwargs)
    472 def FrozenDict(*args, **kwargs) -> Frozen:
--> 473     return Frozen(dict(*args, **kwargs))

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/h5netcdf_.py:231, in <genexpr>(.0)
    229 def get_variables(self):
    230     return FrozenDict(
--> 231         (k, self.open_store_variable(k, v)) for k, v in self.ds.variables.items()
    232     )

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/xarray/backends/h5netcdf_.py:195, in H5NetCDFStore.open_store_variable(self, name, var)
    192 def open_store_variable(self, name, var):
    193     import h5py
--> 195     dimensions = var.dimensions
    196     data = indexing.LazilyIndexedArray(H5NetCDFArrayWrapper(name, self))
    197     attrs = _read_attributes(var)

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/h5netcdf/core.py:252, in BaseVariable.dimensions(self)
    250 """Return variable dimension names."""
    251 if self._dimensions is None:
--> 252     self._dimensions = self._lookup_dimensions()
    253 return self._dimensions

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/h5netcdf/core.py:133, in BaseVariable._lookup_dimensions(self)
    132 def _lookup_dimensions(self):
--> 133     attrs = self._h5ds.attrs
    134     # coordinate variable and dimension, eg. 1D ("time") or 2D string variable
    135     if (
    136         "_Netcdf4Coordinates" in attrs
    137         and attrs.get("CLASS", None) == b"DIMENSION_SCALE"
    138     ):

File ~/opt/anaconda3/envs/install/envs/pangeo-forge/lib/python3.9/site-packages/h5netcdf/core.py:124, in BaseVariable._h5ds(self)
    120 @property
    121 def _h5ds(self):
    122     # Always refer to the root file and store not h5py object
    123     # subclasses:
--> 124     return self._root._h5file[self._h5path]

AttributeError: 'NoneType' object has no attribute '_h5file'



```

</details>

cisaacstern commented on .pre-commit-config.yaml:
```diff
@@ -39,4 +39,4 @@ repos:
     rev: master
     hooks:
-    -   id: rstcheck
+    -   id: rstcheck
```
```suggestion
    -   id: rstcheck

```

cisaacstern replied:
@norlandrhagen, I think adding this newline will get us back to even with `master` for this file

cisaacstern commented:
@norlandrhagen, I just committed to this PR, so you'll have to fetch/rebase from your remote.

Re: the last comment from @pangeo-forge-bot, can you rename `meta.yml` -> `meta.yaml`?

Re: the AttributeError, are these files in NetCDF3 format by chance?

norlandrhagen commented:
According to ncdump, all the files are netcdf4.

cisaacstern commented:
I just tried downloading one of these source files
```
$ wget 'https://knmi-ecad-assets-prd.s3.amazonaws.com/ensembles/data/Grid_0.1deg_reg_ensemble/tg_ens_mean_0.1deg_reg_v23.1e.nc'
```
and opening it with xarray. Which worked. So at least for this file, there are no special kwargs required to open these files locally.

@norlandrhagen, could you try running a local test with the function executor
```python
recipe.copy_pruned().to_function()()
```
?

This will take longer than the Dask executor (which I saw you were using in the traceback above), but it will remove a potential confounding factor from the diagnosis.

norlandrhagen commented:
Yeah, all the files seem to open fine with xarray. I'll try re-running with the `to_function` executor and get some logs.

cisaacstern commented:
@norlandrhagen, I realize we don't have a conclusive diagnosis for the `AttributeError` you mentioned above, but it's cheap enough to run a test of this latest commit, so I'll try that now.

cisaacstern commented:
/run recipe-test recipe_run_id=998

norlandrhagen commented:
Closed in favor of: https://github.com/pangeo-forge/staged-recipes/pull/169
